### ISSUE TITLE
improve the logic of template updates

### ISSUE THREAD

**Issue by rokroskar** [SwissDataScienceCenter/renku-python#2460]
At the moment a project is only migrated to a new version of a template _at the same git ref_ which means that for tags the version of the template is always "current". This makes the template update essentially unusable (templates used in renku deployments are always using tags as refs and the commits they reference never change by definition). 

This leads to bad ux when, for example, a base image has a breaking change that requires a change to a `Dockerfile` and we break the project by updating the base image but not the rest of the `Dockerfile`.

**Comment by mohammad-sdsc** [comment]
There are some other possible issues/improvements:
* We should allow a forced template update which can either update template to a provided version or use a completely new template for the project (e.g. Python -> R). This will be useful for older projects that don't have template metadata set for them and therefore cannot be updated.
* Template update fails if both user and template change a file; this should be allowed for some files (e.g. a `README`) and update shouldn't fail. List of such files comes from the template.
* Instead of completely failing an update (if both parties change a file) we probably should apply possible updates and return the list of files that can't be updated. This can be controlled by a flag.
* Currently, if the update fails some files might have been already updated and the project's repo will remain in a dirty state.
* Checksums of template files should be updated after a template update.

**Comment by rokroskar** [comment]
Thanks for the input @mohammad-sdsc! 

* I like the idea of a "forced" update to transform the project type. This is basically `renku init --force`, but could be packaged slightly differently? What I've wished for in the past there was that I would be prompted _per file_ whether I wanted to overwrite it or not. 
* re: partial updates and merging changes - could we not use git for this? And only fail if git tells us there are conflicts between 
* separate endpoint - you mean other than the one used for general migration?

**Comment by mohammad-sdsc** [comment]
> * I like the idea of a "forced" update to transform the project type. This is basically `renku init --force`, but could be packaged slightly differently? What I've wished for in the past there was that I would be prompted _per file_ whether I wanted to overwrite it or not.

`renku init --force` resets the metadata but a forced template update should preserve project's metadata and only change template files. We have the prompt when initializing a project in an existing directory; we can implement a similar one for the templates.

> * re: partial updates and merging changes - could we not use git for this? And only fail if git tells us there are conflicts between

We can do that for sure. My idea was more like allowing some files to be updated even if there are conflicts in other files (either detected by git or by using checksums). However, it's for sure safer to fail the whole update if there are conflicts so we need to discuss this idea in more details.

> * separate endpoint - you mean other than the one used for general migration?

Right. In CLI, it would be cleaner to have a separate command specially if we want to allow forced template updates. For me, it's confusing to run `migrate` to update the templates. In UI, it's also cleaner to separate metadata migration and template update visually but the API calls are hidden from the users so we still can use the same migration endpoint. But again since the endpoints are still API for the UI team it's also better to have a dedicated endpoint for template updates.

**Comment by rokroskar** [comment]
I agree, it would be much cleaner to have a separate endpoint and command.

**Comment by Panaetius** [comment]
> Template update fails if both user and template change a file; this should be allowed for some files (e.g. a README) and update shouldn't fail. List of such files comes from the template. This functionality can be extended to specify a conflict resolution strategy for each file: ignore (e.g. README.md), merge (e.g. requirements.txt), raise (e.g. .gitlab-ci.yml) and so on.

I believe this is false. If a user modified a file, and it's not considered an immutable_template_file, it is not touched by the update and left as-is, but the template update itself proceeds (and doesn't fail).

I'm not sure merging is a solution. On init we have some dedicated strategies on how to deal with existing files, e.g. .gitignore gets appended to and README.md will be left as is. But I'd say the safe default at the moment is to not touch user modified files. The alternative is the template author specifying what should be done with each file (append, ignore, merge, overwrite). But maybe that's too much complexity.

> Checksums of template files should be updated after a template update.
Wow that's indeed missing :see_no_evil: I must have forgotten to add that.

> Right. In CLI, it would be cleaner to have a separate command specially if we want to allow forced template updates. For me, it's confusing to run migrate to update the templates. In UI, it's also cleaner to separate metadata migration and template update visually but the API calls are hidden from the users so we still can use the same migration endpoint. But again since the endpoints are still API for the UI team it's also better to have a dedicated endpoint for template updates.

There are some dependencies. Rendering the template sets the renku-version in the Dockerfile to the current renku-version, and without accompanying metadata migration, that would potentially leave the project in a state where users can't execute commands unless they then migrate the metadata as well. Doing first template update and then metadata migration as part of one call ensures this doesn't happen. Similarly, if I `renku init --force` with preserving metadata, maybe metadata was on renku version 1.0.0 but the template update causes the Dockerfile to refer to 2.0.0, so a migration of metadata is needed.

On a side note, I believe the Dockerfile update code we have would modify the Dockerfile, which would then prevent the template update from updating the Dockerfile in the future (since it contains local modifications), which should definitely not be the case.


As to the large point. Using git references as versioning was a stopgap implementation that was easy to do and basically works. So just a low hanging fruit. But I never intended it to be set in stone forever, it was more something that "works" until we can come up with some more elaborate versioning for templates. It has other shortcomings as well, like if I use python-minimal and there is a change to r-minimal, the commit still changes and I'd be prompted to do an update that doesn't change anything.

Now instead of shoehorning some forced template update to fix the shortcomings of this approach, I'd rather just implement some proper versioning, not dependent on git (at least not directly) in the manifest.yml. One way would be an ordered list, like 
```
- folder: python-minimal
  name: Basic Python (3.9) Project
  description: The simplest Python-3.9-based renku project with a basic directory structure and necessary supporting files.
  variables: {}
  icon: python-minimal.png
  versions:
    - '1': '36d1ab1b1a2996c9532095043cdd8ce6779c38c0'
    - '2': 'f8ee49bff0a8b97503fad261e87ed51ad590ba04'
    - '2b': 'b486a982aee0439547782de5518bf9eee2b57937'
```

or even

```
- folder: python-minimal
  name: Basic Python (3.9) Project
  description: The simplest Python-3.9-based renku project with a basic directory structure and necessary supporting files.
  variables: {}
  icon: python-minimal.png
  versions:
    - '1': 
      commit: '36d1ab1b1a2996c9532095043cdd8ce6779c38c0'
      changelog: "Added awesome stuff"
    - '2': 
      commit: 'f8ee49bff0a8b97503fad261e87ed51ad590ba04'
      changelog: "Added even more awesome stuff"
    - '2b': 
      commit: 'b486a982aee0439547782de5518bf9eee2b57937'
      changelog: "Frank broke things and this fixes it"
```

Then template maintainers are free to modify templates and mark commit as versions as they please. Because, the tags are ok for our templates, but third party templates aren't linked to tags in a renku release, so wouldn't be supported by what's proposed above.

An extension to the above would be if each version could contain a `previous_version: "2"` entry, then you could have separate chains of versions (like branches) but I think that'd be too much complexity.

We'd still want a `renku init --force` that preserves metadata and overwrites (?) template data, to support projects that were done pre-checksum to post-checksum templates or to maybe switch from python to R templates, as mentioned above.

**Comment by mohammad-sdsc** [comment]
Template update doesn't fully support Semantic Versioning when looking in the Dockerfile for a renku version: https://github.com/SwissDataScienceCenter/renku-python/blob/2879c55314bbee58c040a455fe1882136e404595/renku/core/management/migrate.py#L327 Therefore, it's not possible to update projects that use a dev version of renku (e.g. `1.0.0.dev1+g942be48d`).

**Comment by Panaetius** [comment]
That was by design, the reasoning was that our released service versions should never be a dev version. So if a user has a dev version in their Dockerfile, it's because they put it there manually. If they put it there manually, they probably had good reason to use a dev version and they don't want the service to automatically update that.

**Comment by mohammad-sdsc** [comment]
This happened to Sekhar when he deployed an env with a dev version of renku and he updated the renku version via UI; check for renku version could not be done after the update because of the dev version in the Dockerfile. I don't see any harm to fully support SemVer to enable this use case.

**Comment by Panaetius** [comment]
I don't have a strong opinion either way. I guess it makes testing easier and 99.9% of our users never end up with a dev version.

I remember there was also an issue where before a dev release after 0.11.0 would be 0.11.1.dev1, and then when we actually release 0.11.1, it would be considered lower than 0.11.1.dev1 and wouldn't get upgraded to. But I think that hasn't been the case for a while

**Comment by mohammad-sdsc** [comment]
I've added a `renku template` command to the CLI. Its subcommands include `ls` and `show` to list templates and show detailed info for a template which replace `--list-templates` and `--describe` from the `renku init` command. There is also a `set` subcommands that allows setting a template for a project, and an `update` subcommand to update a project's template. By default `set` overwrites existing files in the project (similar to `init`) while `update` does not overwrite them. They both accept an `--interactive` flag which asks users what to do with each file. 

Updating or setting project's template preserves the `RENKU_VERSION` set in the Dockerfile even if the Dockerfile is overwritten.

As for the template versioning, I've decided to keep what we currently do. There are three variables that define a specific template version: `template_source`, `template_ref`, and `template_version`. For Renku templates, `template_source` is `"renku"`, `template_ref` is `None`, and `template_version` is the Renku CLI version. For templates from other sources, `template_source` is the template repo's URL, `template_ref` is either `None` or a Git reference (branch, tag, SHA, ...), and `template_version` is the SHA of the reference.

If the `template_ref` is a tag or SHA, the template won't be updated since `template_version` is always the same. The `template update` command includes a `--force` flags that makes it to ignore the reference and update the template anyways. I also removed the `automated_template_update` field from templates which would prevent updating a template since we can achieve the same by setting a `template_ref`.

All these changes are in the CLI; the service interface is the same as before. I've noticed that when we create a project from the UI from a Renku template, the template is set as an external template (i.e. `template_source` is set to Renku template repo URL and `template_ref` is the tag). We should probably change this to set the `template_source` to be `"renku"` for easier template update (i.e. without `--force`). I'll discuss this with the UI team once the CLI part is done.
